docs/development/pigs/pig-028.rst

```diff
@@ -0,0 +1,285 @@
+        self.frac_remain = frac_remain
+
+    @staticmethod
+    def _parse_datasets(datasets):
+        from gammapy.datasets import Dataset, Datasets
+
+        if isinstance(datasets, (list, Dataset)):
+            datasets = Datasets(datasets)
+        return datasets, datasets.parameters.free_parameters
+
+    @staticmethod
+    def _update_models(models, results):
+        posterior = results["posterior"]
+        samples = results["samples"]
+        for i, par in enumerate(models.parameters.free_parameters):
+            par.value = posterior["mean"][i]
+            par.error = posterior["stdev"][i]
+            par._samples = samples[:,i] * par.unit
+
+        # Compute a covariance from the samples?
+        covariance = Covariance(models.parameters.free_parameters, np.cov(samples.T))
+        models.covariance=covariance
+        return models
+    
+    def run(self, datasets):
+        datasets, parameters = self._parse_datasets(datasets=datasets)
+        parameters = parameters.free_parameters
+
+        # create log likelihood function
+        # need to remove prior penalty
+        like = SamplerLikelihood(function=datasets.stat_sum_no_prior, parameters=parameters)
+        
+        # create ultranest object
+        self._sampler = ultranest.ReactiveNestedSampler(parameters.names, like.fcn, transform=parameters.prior_inverse_cdf)
+        
+        results = self._sampler.run(min_num_live_points=self.min_num_live_points, frac_remain=self.frac_remain)
+
+        updated_models = self._update_models(datasets.models.copy(), results)
+        
+        return BayesianInference(models=updated_models, results=results)
+
+
+            
+
+
+
+.. code:: python
+
+  class SamplerLikelihood:
+    """Wrapper of the likelihood function used by the sampler.
+
+    Parameters
+    ----------
+    parameters : `~gammapy.modeling.Parameters`
+        Parameters with starting values.
+    function : callable
+        Likelihood function.
+    """
+
+    def __init__(self, function, parameters):
+        self.function = function
+        self.parameters = parameters
+
+    def fcn(self, value):
+        self.parameters.value = value
+        total_stat = -0.5*self.function()
+
+        return total_stat
+
+
+Serialisation of the samples
+-------------
+In UltraNest, a serialisation is proposed to save to disk the samples and results dictionnary. So this can be used directly to `save` and there is also a `read_file` function.
+
+
+Task list 
+-----------------------
+
+The task will be split between different subcategories
+
+*Modifications to existing code*
+1. Add a stat_sum function `stat_sum_no_prior` which returns only the likelihood term Without the penalty term
+2. Add inverse_cdf attributes to the all Priors
+3. Add LogUniform prior (crucial for amplitude)
+
+*Design & decisions of a new Bayesian class*
+1. Should the class be called NestedSampler (only NS) or BayesianFit (would host MCMC and NS)
```

Comment:
this has been a point of the discussion indeed. MCMC are of course more commonly understood & accepted.
However they require some hyper-parameters that are not easy to tune. 
But it's true that we could make the class general enough so that an MCMC sampler could be plugged easily. 
There are some new generation samplers that are much better than `emcee` and that solve part of the above issues.
+1 for `BayesianFit` name